Add callbacks for ompt progress and add test that shows it works

cmcantalupo opened the pull request:
- The ompt callback "dispatch" does not seem to be active.
- The ompt work callback does seem to be active
- [x] TODO: remove dispatch callback until it can be tested
- [x] TODO: validate that the work callback is doing exactly what is expected
- [x] TODO: Add loose assertions about total runtime of the test_progress run
- [x] TODO: Add assertions that the progress signal goes up twice
- [x] TODO: Add assertions that the progress signal is monotone within each OMPT detected region
- [x] TODO: Clean up history

cmcantalupo committed:
Add a helper to integration.test.until to find test binaries

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add integration test for progress reporting

- A couple of clean up issues with test_hint_time
  + Add to __main__ and remove unused member

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Enable use of ompt callback for geopm_tprof_init()

- This requires that geopm_tprof_init() have the expectation that one thread
  per process will call it with the total amount of work per thread.
- To support this requirement the interface to Profile::init_thread() has changed.
- TODO: Fix unit tests for Profile::init_thread() use.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Do not use C++ Profile interface outside of DefaultProfile

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix Profile unit tests to reflect change in the way tprof_init works

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Disable AgentFactory test for ee agent creation

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented on integration/test/test_progress.py:
```diff
@@ -0,0 +1,286 @@
+import pandas
+import numpy
+import socket
+import matplotlib.pyplot as plt
```
unused import

cmcantalupo committed:
Add documentation to progress integration test

- Also added a few assertions.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix up monotone check

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix race condition with previous solution

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dianarg commented on integration/test/test_progress.py:
```diff
@@ -0,0 +1,379 @@
+        geopmpy.error.exc_clear()
+
+        # Create machine
+        cls._machine = machine.Machine()
+        try:
+            cls._machine.load()
+            sys.stderr.write('Warning: {}: using existing file "machine.json", delete if invalid\n'.format(cls._test_name))
+        except RuntimeError:
+            cls._machine.save()
```
could be a helper, if this is going to be a pattern in the tests.  I seem to remember we didn't want people to call the Machine constructor directly.

cmcantalupo replied:
I will move it into a machine method, "Machine.try_load_or_save()" or something like that.  I'm not sure I recall any issues with calling machine.Machine(), but I do think the Machine interface could probably use an overhaul.  It is a little more surprising than I'd like.

cmcantalupo replied:
Your memory is correct, I added a helper function "Machine.try_machine()", in other cases we use "Machine.get_machine()".

dianarg commented on integration/test/test_progress.py:
```diff
@@ -0,0 +1,379 @@
+        self.assertLessEqual(0, min(progress.diff()[1:]), err_msg)
+
+    def test_linear_progress_triad(self):
+        """Test that the triad region with with the post reports progress that
```
extra with

dianarg replied:
post -> post()

cmcantalupo replied:
Thanks!

dianarg commented on integration/test/test_progress.py:
```diff
@@ -0,0 +1,379 @@
+        triad_post_hash = self.get_hash('triad_with_post')
+        triad_post_df = grouped_df.get_group(triad_post_hash)
+        total_progress = self.sum_progress(cpu_with_progress, triad_post_df)
+        err_msg = 'Bad fit for triad total proggress'
```
proggress -> progress, more below

cmcantalupo replied:
Thank you!

dianarg commented on integration/test/test_progress.py:
```diff
@@ -0,0 +1,379 @@
+        triad_post_df = grouped_df.get_group(triad_post_hash)
+        total_progress = self.sum_progress(cpu_with_progress, triad_post_df)
+        err_msg = 'Bad fit for triad total proggress'
+        self.check_progress(triad_post_df['TIME'],
```
check_monotone() should also be valid for total progress

cmcantalupo replied:
I will try to bring that back.

cmcantalupo replied:
The problem with the total progress analysis is that I had to select based on the region hash for the entire board (i.e. the synchronous part).  This meant that the end of the lagging ranks progress was cut off by the group_by() selection.  This caused issues with the helper check assertsions.  The sum of progress is a little contrived, and checking each of them is a finer grained check (now selecting on per-cpu region hash).

dianarg commented on integration/test/util.py:
```diff
@@ -142,6 +142,24 @@ def get_config_value(key):
+    path_list.append(os.path.join(int_dir, 'build/integration/test/.libs'))
+    bin_list = [os.path.join(pp, bin_name) for pp in path_list]
+    for bb in bin_list:
+        if os.path.exists(bb):
```
maybe also check isfile(), os.access() here if the errors from Popen are not as useful

cmcantalupo replied:
Great, I'll look into that.

dianarg commented on src/DefaultProfile.cpp:
```diff
@@ -183,15 +183,18 @@ extern "C"
-            catch (...) {
-                err = geopm::exception_handler(std::current_exception());
+            if (g_pmpi_prof_enabled) {
+                prof.thread_work(num_work_unit);
```
Don't the non-lead threads pass 0 here?

cmcantalupo replied:
Only the lead thread has "g_pmpi_prof_enabled", all the threads have "g_pmpi_tprof_enabled".  I settled on this solution so that the lead thread is responsible for setting the number of work units, but each thread is responsible for zeroing its own counter.  This avoids an error where the lead thread would zero the counter after the non-lead thread had already started incrementing it.  There should be a comment here explaining that!

cmcantalupo committed:
Fix it better

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on integration/test/test_progress.py:
```diff
@@ -0,0 +1,379 @@
+        # Set the job size parameters
+        cls._num_node = 1
+        cls._num_rank = 10
+        cls._num_active_cpu = cls._machine.num_core() - 2 * cls._machine.num_package()
```
Speaking of the Machine object, it would be great if we could ask it for the number of active CPUs to avoid sprinkling the magic ```2``` around all the tests.  This will be nice if/when we change where and how the Controller runs (application mode).

cmcantalupo replied:
The unit test has been failing since we started deriving the hint from the ApplicationStatus lock free solution and not based on messages in the ApplicationRecordLog.  This unit test needs to be updated to mock the ApplicationStatus object, not the ApplicationRecordLog object.

In terms of the magic number 2, this is based on the geopmpy.launch behavior (reserve one core for the OS and one core for GEOPM, which is what this number two reprsents).  The machine file does not contain information about the launch command was issued, or what the application cpu affinities are.

bgeltz reviewed (changes requested):
I'm seeing a unit test failure with this PR when I build with debug enabled (release too):
```
FAIL: test/gtest_links/ApplicationSamplerTest.hint_time
```

The log shows:
```
Note: Google Test filter = ApplicationSamplerTest.hint_time
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from ApplicationSamplerTest
[ RUN      ] ApplicationSamplerTest.hint_time
test/ApplicationSamplerTest.cpp:387: Failure
Expected equality of these values:
  0.0
    Which is: 0
  compute_time
    Which is: nan
test/ApplicationSamplerTest.cpp:389: Failure
Expected equality of these values:
  0.0
    Which is: 0
  network_time
    Which is: nan
test/ApplicationSamplerTest.cpp:391: Failure
Expected equality of these values:
  0.0
    Which is: 0
  memory_time
    Which is: nan
test/ApplicationSamplerTest.cpp:393: Failure
Expected equality of these values:
  0.0
    Which is: 0
  compute_time
    Which is: nan
test/ApplicationSamplerTest.cpp:395: Failure
Expected equality of these values:
  0.0
    Which is: 0
  network_time
    Which is: nan
test/ApplicationSamplerTest.cpp:397: Failure
Expected equality of these values:
  0.0
    Which is: 0
  memory_time
    Which is: nan

GMOCK WARNING:
Uninteresting mock function call - returning directly.
    Function call: check_sample_end()
NOTE: You can safely ignore the above warning unless this call should not happen.  Do not suppress it by blindly adding an EXPECT_CALL() if you don't mean to enforce the call.  See https://github.com/google/googletest/blob/master/googlemock/docs/CookBook.md#knowing-when-to-expect for details.
test/ApplicationSamplerTest.cpp:411: Failure
Expected equality of these values:
  0.0
    Which is: 0
  compute_time
    Which is: nan
test/ApplicationSamplerTest.cpp:413: Failure
Expected equality of these values:
  0.0
    Which is: 0
  network_time
    Which is: nan
test/ApplicationSamplerTest.cpp:415: Failure
Expected equality of these values:
  0.0
    Which is: 0
  memory_time
    Which is: nan
test/ApplicationSamplerTest.cpp:417: Failure
Expected equality of these values:
  0.0
    Which is: 0
  compute_time
    Which is: nan
test/ApplicationSamplerTest.cpp:419: Failure
Expected equality of these values:
  0.0
    Which is: 0
  network_time
    Which is: nan
test/ApplicationSamplerTest.cpp:421: Failure
Expected equality of these values:
  0.0
    Which is: 0
  memory_time
    Which is: nan

GMOCK WARNING:
Uninteresting mock function call - returning directly.
    Function call: check_sample_end()
NOTE: You can safely ignore the above warning unless this call should not happen.  Do not suppress it by blindly adding an EXPECT_CALL() if you don't mean to enforce the call.  See https://github.com/google/googletest/blob/master/googlemock/docs/CookBook.md#knowing-when-to-expect for details.
test/ApplicationSamplerTest.cpp:436: Failure
Expected equality of these values:
  0.0
    Which is: 0
  compute_time
    Which is: nan
test/ApplicationSamplerTest.cpp:438: Failure
Expected equality of these values:
  1.0
    Which is: 1
  network_time
    Which is: nan
test/ApplicationSamplerTest.cpp:440: Failure
Expected equality of these values:
  0.0
    Which is: 0
  memory_time
    Which is: nan
test/ApplicationSamplerTest.cpp:442: Failure
Expected equality of these values:
  1.0
    Which is: 1
  compute_time
    Which is: nan
test/ApplicationSamplerTest.cpp:444: Failure
Expected equality of these values:
  0.0
    Which is: 0
  network_time
    Which is: nan
test/ApplicationSamplerTest.cpp:446: Failure
Expected equality of these values:
  0.0
    Which is: 0
  memory_time
    Which is: nan

GMOCK WARNING:
Uninteresting mock function call - returning directly.
    Function call: check_sample_end()
NOTE: You can safely ignore the above warning unless this call should not happen.  Do not suppress it by blindly adding an EXPECT_CALL() if you don't mean to enforce the call.  See https://github.com/google/googletest/blob/master/googlemock/docs/CookBook.md#knowing-when-to-expect for details.
test/ApplicationSamplerTest.cpp:461: Failure
Expected equality of these values:
  0.0
    Which is: 0
  compute_time
    Which is: nan
test/ApplicationSamplerTest.cpp:463: Failure
Expected equality of these values:
  3.0
    Which is: 3
  network_time
    Which is: nan
test/ApplicationSamplerTest.cpp:465: Failure
Expected equality of these values:
  0.0
    Which is: 0
  memory_time
    Which is: nan
test/ApplicationSamplerTest.cpp:467: Failure
Expected equality of these values:
  1.0
    Which is: 1
  compute_time
    Which is: nan
test/ApplicationSamplerTest.cpp:469: Failure
Expected equality of these values:
  0.0
    Which is: 0
  network_time
    Which is: nan
test/ApplicationSamplerTest.cpp:471: Failure
Expected equality of these values:
  2.0
    Which is: 2
  memory_time
    Which is: nan

GMOCK WARNING:
Uninteresting mock function call - returning directly.
    Function call: check_sample_end()
NOTE: You can safely ignore the above warning unless this call should not happen.  Do not suppress it by blindly adding an EXPECT_CALL() if you don't mean to enforce the call.  See https://github.com/google/googletest/blob/master/googlemock/docs/CookBook.md#knowing-when-to-expect for details.
test/ApplicationSamplerTest.cpp:486: Failure
Expected equality of these values:
  3.0
    Which is: 3
  compute_time
    Which is: nan
test/ApplicationSamplerTest.cpp:488: Failure
Expected equality of these values:
  3.0
    Which is: 3
  network_time
    Which is: nan
test/ApplicationSamplerTest.cpp:490: Failure
Expected equality of these values:
  0.0
    Which is: 0
  memory_time
    Which is: nan
test/ApplicationSamplerTest.cpp:492: Failure
Expected equality of these values:
  1.0
    Which is: 1
  compute_time
    Which is: nan
test/ApplicationSamplerTest.cpp:494: Failure
Expected equality of these values:
  3.0
    Which is: 3
  network_time
    Which is: nan
test/ApplicationSamplerTest.cpp:496: Failure
Expected equality of these values:
  2.0
    Which is: 2
  memory_time
    Which is: nan
[  FAILED  ] ApplicationSamplerTest.hint_time (2 ms)
[----------] 1 test from ApplicationSamplerTest (2 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (3 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] ApplicationSamplerTest.hint_time

 1 FAILED TEST
```

cmcantalupo committed:
Fix up test based on last change

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on integration/test/test_progress.py:
```diff
@@ -0,0 +1,379 @@
+        self.check_overhead('dgemm', 0.05)
+
+    def test_overhead_triad(self):
+        """Test that the overhead not huge for triad which does only 3
+        operations per work chunk (FMA,add,store).
```
```suggestion
        """Test that the overhead is not huge for triad which does only 3
        operations per work chunk (FMA, add, and store).
```

bgeltz commented on src/ApplicationStatus.cpp:
```diff
@@ -76,7 +76,6 @@ namespace geopm
         // initialize shmem if all zero is not appropriate
         for (int cpu = 0; cpu < m_num_cpu; ++cpu) {
             set_process({cpu}, -1);
-            set_total_work_units(cpu, 0);
```
This seems important.  Why is it not necessary?

bgeltz commented on src/Profile.cpp:
```diff
@@ -498,7 +495,8 @@ namespace geopm
                 // region.
-                m_app_status->set_total_work_units(cpu, 0);
+                m_app_status->set_hash(cpu, m_current_hash);
+                m_app_status->reset_work_units(cpu);
```
I expected to see someone setting the work units to 0 at the first region entry, but I'm not seeing it.

cmcantalupo committed:
Get rid of test that sums progress

- Since we were grouping by the REGION_HASH column, which is
  aggregated over the board, we were missing all asychonous samples.
- This causes an arbitrary cut off of the end of the data.
- To avoid this, now just select each cpu progress independently.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
We are only using progress when total_work is non-zero

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix another race condition and update the test

- Test now allows progress to reach the maximum value
  and stay there when doing linear fit.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix issue with ompt_callback

- Work executed in parallel regions that is not part of a
  work share seems to call the work callback with count == 1.
- Fix test for max value (don't remove the value before
  testing it is there)

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Address review feedback

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix signed / unsigned comparison

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

alawibaba commented on integration/test/test_progress.py:
```diff
@@ -0,0 +1,349 @@
+        self.assertLess(len(progress) * 0.6, sum(good_idx))
+        progress = progress[good_idx]
+        time = time[good_idx]
+        poly_out = numpy.polyfit(time,
```
This is fine, but if we have a performance issue in the future there is repeated work in this call that we could eliminate if this is slow.

cmcantalupo replied:
The test runs very quickly, so for now this is not a problem.  If we start using this pattern in tests that use more data, we may want to look into performance improvements.

alawibaba commented:
Hey LGTM. I had one little comment about one of the tests but it's just a comment.
